drivers: modem: Sim7080 modem driver refactor

lgehreke opened the pull request:
The sim 7080 driver has a lot functionality crammed into one source file. This makes it hard to maintain. Furthermore some bugs were fixed during the refactor.

Changes:
* Split modem driver into multiple files
* One file per modem functionality
* Verifying xtra file validity on aided gps
* Modem driver uses fixed baudrate to evade autobaud sequence

Additions:
* Added multiple boot modes
* Added force reset function
* Added function to query time
* Added battery measurement function
* Added function to query iccid
* Added gpio set function

Bugfixes:
* Corrected socket connect behavior
* Fixed wrong pdp context on offload_connect
* Send sequence queries the available tx space before sending

lgehreke committed:
drivers: sim7080: Own simcom modem driver directory

Placed sim7080 driver in separate directory and split it in multiple
source files for better readability.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: Using fixed baudrate for sim7080

On first boot a fixed baudrate is set for the sim7080. This makes power
on detection more reliable.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: Added modem model check to sim7080 driver

Check the modem model on boot.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: own file for pdp handling

PDP context handling takes place in separate source file for better
readability.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: Implemented boot modes for sim7080

Modem can either attach to network after initialization or be turned
off.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: added stop functions for network and gnss

Network and gnss can be disabled with stop functions instead of power
cycling the modem. The start functions will also not power cycle the
modem. In order to call the start functions the modem needs to be
booted.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: moved sim7080 socket related urc handling to socket file

Moved socket related urc handling to dedicated file.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: corrected sim7080 socket error codes.

Socket functions return EINVAL instead of EAGAIN when modem is
not in networking mode. Using EAGAIN could lead to a infinite sleep
loop in offload_sendmsg().

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: removed sleep from offload_sendmsg

Removed sleep to prevent infinite loop in sendmsg.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: fixed offload_connect for sim7080

Connection id and pdp index were swapped. Corrected this and
hardcoded pdp index since 0 is always used as index.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: Corrected socket send sequence.

Query available tx size before sending data instead of using
hardcoded number. Removed unnecessary strg+z send after data.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: fixed socket connect behavior

The modem handler error can not be used to transport the result
of CAOPEN since it is overwritten by the trailing OK.
Socket does not get closed forcefully when connecting failed.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: socket can be closed if not connected

Removed connected check from offload_close in order to be able
to close non-connected sockets.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: preventing recvfrom hangup on closed socket

If a socket gets closed by the network while waiting on data
in offload_recvfrom this would hangup forever since modem_socket_wait_data
does not support timeouts. When the socket gets closed this wait is
unblocked.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: added gpio set function

Added functionality to set modem gpio pins.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: added function to get iccid

Added function to get the iccid number.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: Added battery measurement function for sim7080

Added function to measure battery voltage, battery charge status
and battery connection level.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: added function to query the modem state.

Added function to query the modem state from application side.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: added command to inquire ue system information

Added command to query ue system information from the modem.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: sim7080: implemented gnss xtra functionality.

Added funtions to download a gnss xtra file, query its validity
and use it in gps.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: added function to query local time

Added funtion to query local time and added injection time to
gnss xtra validity query function.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: querying xtra validity during gnss start

Starting gnss with xtra functionality is only possible if the
validity of the xtra file was queried. Enabling xtra will be skipped
if the file is not valid

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: added force reset function.

Added function to forcefully reset the modem by holding the pwrkey
for 15 seconds.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: Allwing ftp when networking is already active

Added check that makes sure ftp works when the network context is
already active.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

rerickson1 reviewed (dismissed):
Do a rebase and see if CI is fixed.

lgehreke committed:
drivers: modem: Fixed style issues of sim7080 driver

Fixed zephyr style violations.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

rerickson1 commented:
Please fix compliance issues.

lgehreke commented:
> Please fix compliance issues.

Is this a bug in the compliance checks? Why should there be spaces around the ':' of a goto label? This is used in multiple occasions but the pipeline fails only at this line.

bjarki-andreasen commented:
> > Please fix compliance issues.
> 
> Is this a bug in the compliance checks? Why should there be spaces around the ':' of a goto label? This is used in multiple occasions but the pipeline fails only at this line.

it seems to be a bug since the label "error" is used, just use another label like out or something

rerickson1 commented:
> > Please fix compliance issues.
> 
> Is this a bug in the compliance checks? Why should there be spaces around the ':' of a goto label? This is used in multiple occasions but the pipeline fails only at this line.

Also, fix trailing spaces.

lgehreke committed:
drivers: modem: sim7080: removed error label in modem_setup

Removed label because the compliance checks fail.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke commented:
> > > Please fix compliance issues.
> > 
> > 
> > Is this a bug in the compliance checks? Why should there be spaces around the ':' of a goto label? This is used in multiple occasions but the pipeline fails only at this line.
> 
> Also, fix trailing spaces.

Fixed in this commit f7cc080e2637e834009422fb56f43732c332e816 but pipeline still complains. Any thoughts on that?

lgehreke commented:
> > > > > Please fix compliance issues.
> > > > 
> > > > 
> > > > Is this a bug in the compliance checks? Why should there be spaces around the ':' of a goto label? This is used in multiple occasions but the pipeline fails only at this line.
> > > 
> > > 
> > > Also, fix trailing spaces.
> > 
> > 
> > Fixed in this commit [f7cc080](https://github.com/zephyrproject-rtos/zephyr/commit/f7cc080e2637e834009422fb56f43732c332e816) but pipeline still complains. Any thoughts on that?
> 
> trailing whitespace are tabs and spaces following trailing lines added, you should be able to enable whitespace rendering in your IDE so you can see them, I added a traling space below to show what its complaining about.
> 
> ```
> int my_line = 0 
> ```
> 
> IDEs like vscode have a "trim trailing whitespace" option which will do this for you when you save the file

Running `git rebase main --whitespace=fix` resolved the issue

lgehreke committed:
drivers: modem: sim7080: Added missing doxygen docu to public header.

Added missing doxygen comments to public header.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

lgehreke committed:
drivers: modem: sim7080: Made dns timeout and retry configurable

Timeout and retries for DNS lookups were hardcoded. This commit
introduces kconfig settings for the default values and functions
for runtime configuration.

Signed-off-by: Lukas Gehreke <lk.gehreke@gmail.com>

kartben reviewed (approved):
+1 for docs, thanks for your efforts on this, much appreciated!
